Classes/common/Encryption/Keychain/CDTEncryptionKeychainManager.m

```diff
@@ -0,0 +1,193 @@
+//
+//  CDTEncryptionKeychainManager.m
+//
+//
+//  Created by Enrique de la Torre Fernandez on 09/04/2015.
+//
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//  http://www.apache.org/licenses/LICENSE-2.0
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+//
+
+#import <CommonCrypto/CommonCryptor.h>
+
+#import "CDTEncryptionKeychainManager.h"
+
+#import "CDTEncryptionKeychainUtils.h"
+#import "CDTEncryptionKeychainConstants.h"
+
+#import "CDTLogging.h"
+
+@interface CDTEncryptionKeychainManager ()
+
+@property (strong, nonatomic, readonly) CDTEncryptionKeychainStorage *storage;
+
+@end
+
+@implementation CDTEncryptionKeychainManager
+
+#pragma mark - Init object
+- (instancetype)init { return [self initWithStorage:nil]; }
+
+- (instancetype)initWithStorage:(CDTEncryptionKeychainStorage *)storage
+{
+    self = [super init];
+    if (self) {
+        if (storage) {
+            _storage = storage;
+        } else {
+            CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"Storage is mandatory");
+
+            self = nil;
+        }
+    }
+
+    return self;
+}
+
+#pragma mark - Public methods
+- (NSData *)retrieveEncryptionKeyDataUsingPassword:(NSString *)password
+{
+    CDTEncryptionKeychainData *data = [self.storage encryptionKeyData];
+    if (!data || ![self validatedEncryptionKeyData:data]) {
+        return nil;
+    }
+    
+    NSData *aesKey =
+        [self generateAESKeyUsingPBKDF2ToDerivePassword:password
+                                               withSalt:data.salt
+                                             iterations:data.iterations
+                                                 length:CDTENCRYPTION_KEYCHAIN_AES_KEY_SIZE];
+
+    NSData *dpk = [self decryptCipheredDpk:data.encryptedDPK usingAESWithKey:aesKey iv:data.iv];
+
+    return dpk;
+}
+
+- (NSData *)generateEncryptionKeyDataUsingPassword:(NSString *)password
+{
+    NSData *dpk = nil;
+
+    if (![self encryptionKeyDataAlreadyGenerated]) {
+        dpk = [self generateDpk];
+
+        CDTEncryptionKeychainData *keychainData =
+            [self keychainDataToStoreDpk:dpk encryptedWithPassword:password];
+
+        if (![self.storage saveEncryptionKeyData:keychainData]) {
+            dpk = nil;
+        }
+    }
+
+    return dpk;
+}
+
+- (BOOL)encryptionKeyDataAlreadyGenerated { return [self.storage encryptionKeyDataExists]; }
```

Comment:
I think the method naming of "generate" is misleading because the method also stores the keychain data, in addition to generating it. Is there a reason not to mirror the underlying methods like so:

```
-loadEncryptionKeyDataUsingPassword:
-saveEncryptionKeyDataUsingPassword:
-encryptionKeyDataExists
```

I guess the generate speaks to the purpose of the class, "Use this class to generate a Data Protection Key (DPK)"?